Design hibernate regions (#23)
tikv/rfcs: text/2019-03-04-hibernate-raft.md

broadcast heartbeats to all followers in a long period.

Followers can't know whether there is further write, so it can just consider
the group Idle when it contains logs from current term and has a leader. And
then let leader to drive the log replication. And even followers are in Idle
state, it should also check whether heartbeats arrives in long enough time. If
not, it means the follower may have been removed from the group or leader is
down. It should become Chaos and campaign.

### RaftLogGc

Because raft log is done by proposal, so only leader needs to tick. And when
raft tick is stopped, this tick can be stopped too. When new write proposal
arrives or a new leader is elected, the tick can be started again.

### SplitRegionCheck

This can be handled using the same policy as `RaftLogGc` except that when
rocksdb finishes compaction or merge/split is finished, it needs to tick again
at least one time.

### PdHeartbeat

This is also similar to `RaftLogGc`. However, `PdHeartbeat` must tick when
there are any read proposals to let pd rebalance correctly.

### CheckMerge

It has been implemented in the lazy way already that it's only scheduled when

> a `PreMerge` is proposed while `CommitMerge` is not applied.

Reviewer: What is `PreMerge`?

### CheckPeerStaleState

This tick can't be removed as there is no reliable way to remove stale peer.

Besides leader has to broadcast heartbeats to all followers in a long period
as memtioned above, so this tick can be also used for broadcasting and
checking heartbeats. To make sure followers check the leader heartbeat
reliablely, we can add a intermediate state to group state as PreChaos. If a
follower is Idle, then change it to PreChaos; if it's PreChaos, then change it
to Chaos. So it takes two ticks for follower to actually start campaign while
it only takes one tick for leader to broadcast heartbeats. So followers should
generally stay calm if network and clock is good.

## Drawbacks

> When a leader is down, it may take a longer time to recover than before.

Reviewer: What about transit to `Chaos` when a follower receives a client request?

Author: It's part of the design, see L44. But client still needs to wait at least a round of election.

# Hibernate Region

## Summary

This RFC proposes to hibernate a peer when it reaches a stable state.

## Motivation

There are several ticks when driving a peer. The cost of tick is usually
small. However, when there are a lot of regions in one TiKV instances, the

> cost can't be eligible. We have observed that an idle cluster with many

Reviewer 2: eligible -> negligible

regions can waste a lot of CPUs to send useless heartbeats and tick to routine
checks. Even in an online cluster with tens of thousands of regions, enlarge
the ticks' interval can usually lead to better performance. Threaded raftstore
can ease the problem but can't solve it. In general, not all regions keep

> working all the time. Imbalanced load is a very common use case in practice.

Reviewer 3: ```suggestion
working all the time. Unbalanced load is a very common use case in practice.
```

up because of 5. Note that in such case, new leader will not stop ticking even
it is elected and apply the new entry because there is still one out of service
node, which is the old leader. Rule 4 can be more aggressive like if it detects
that it has not received messages from leader for more than election_timeout,
it can start campaign immediately.

If during that time, follower is out of service and recovered before any
proposals, it will start ticking because of 3. To prevent it from disturbing
the leader, PreVote needs to be enabled. And leader will eventually be woken
up and broadcasting heartbeat until all followers have been calmed down and go
to sleep as 7 indicates. If follower is recovered after proposals, leader has
been woken up already, and the whole cluster will be made sure to be up to
date.

If during that time, a new node is added to the cluster, because of 1, leader
will keep ticking until the conf change is complete and new node has caught up
all the logs. If a node is removed, and the node is isolated, we need another
mechanism to make it removed. This will be discussed when talking about
`CheckPeerStaleState`.

### RaftLogGc

> Apparently, only leader needs to tick. And when raft tick is stopped, this

Reviewer 3: ```suggestion
Only leader needs to tick. And when raft tick is stopped, this
```

Let's make sure the RFC isn't working on assumptions. Is it true that only the leader needs to tick or not?

tick can be stopped too. When new write proposal arrives or a new leader
is elected, the tick can be started again.

### SplitRegionCheck

This can be handled using the same policy as `RaftLogGc` except that when
rocksdb finishes compaction or merge/split is finished, it needs to tick again
at least one time.

### PdHeartbeat

This is also similar to `RaftLogGc`. However, `PdHeartbeat` must tick when
there are any read proposals to let pd rebalance correctly.

### CheckMerge

It has been implemented in the lazy way already that it's only scheduled when
a `PreMerge` is proposed while `CommitMerge` is not applied.

### CheckPeerStaleState

Hibernation means it should not be interrupted only when it's really

> necessary. This tick is deprecated completely. We have two ways to clean up

Reviewer 3: Is it worthwhile to investigate removing it completely?

When both leader and followers are hibernated, a proposal arrives. Leader will
start ticking again because of 4. And followers will tick. Note that leader
must tick in this case so that it can keep retry sending out entries when
network is in bad condition. And after the proposal is committed on all peers
and applied on leader, leader will stop ticking because of 1.

If during that time, leader is out of service, client may try to redirect
requests to followers. Because of 4, followers will start ticking. If client
doesn't redirect the requests, followers can still start ticking due to 8.
They start campaigns in the end. And their campaign messages will wake the
whole cluster up because of 5. Note that in such case, new leader will not
stop ticking even it is elected and applies the new entry because there is
still one out of service node, which is the old leader. Rule 4 and 8 can be
more aggressive like if it detects that it has not received messages from
leader for more than election_timeout, it can start campaign immediately.

If during that time, follower is out of service and recovered before any
proposals, it will start ticking because of 3. To prevent it from disturbing

> the leader, PreVote needs to be enabled. And leader will eventually be woken

Reviewer 3: ```suggestion
the leader, PreVote needs to be enabled. This way the leader will eventually be woken
```

up and broadcasting heartbeat until all followers have been calmed down and go
to sleep as 7 indicates. If follower is recovered after proposals, leader has
been woken up already, and the whole cluster will be made sure to be up to
date.

> If during that time, a new node is added to the cluster, because of 1, leader

Reviewer 3: ```suggestion
During this time, if a new node is added to the cluster, because of 1, leader
```
